lib/Controller/MessageApiController.php

```diff
@@ -0,0 +1,171 @@
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCA\Mail\Controller;
+
+use OCA\Mail\Db\LocalMessage;
+use OCA\Mail\Exception\ClientException;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Exception\UploadException;
+use OCA\Mail\Service\AccountService;
+use OCA\Mail\Service\AliasesService;
+use OCA\Mail\Service\Attachment\AttachmentService;
+use OCA\Mail\Service\Attachment\UploadedFile;
+use OCA\Mail\Service\OutboxService;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\Attribute\NoAdminRequired;
+use OCP\AppFramework\Http\Attribute\NoCSRFRequired;
+use OCP\AppFramework\Http\Attribute\UserRateLimit;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\OCSController;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\DB\Exception;
+use OCP\IRequest;
+use Psr\Log\LoggerInterface;
+
+class MessageApiController extends OCSController {
+
+	private ?string $userId;
+
+	public function __construct(
+		string $appName,
+		?string $userId,
+		IRequest $request,
+		private AccountService $accountService,
+		private AliasesService $aliasesService,
+		private AttachmentService $attachmentService,
+		private OutboxService $outboxService,
+		private LoggerInterface $logger,
+		private ITimeFactory $time,
+	) {
+		parent::__construct($appName, $request);
+		$this->userId = $userId;
+	}
+
+	#[UserRateLimit(limit: 5, period: 100)]
+	#[NoAdminRequired]
+	#[NoCSRFRequired]
+	public function send(
+		int $accountId,
+		string $fromEmail,
+		string $subject,
+		string $body,
+		bool $isHtml,
+		array $to,
+		array $cc = [],
+		array $bcc = [],
+		array $references = [],
+	): DataResponse {
+		if($this->userId === null) {
+			return new DataResponse('Account not found.', Http::STATUS_NOT_FOUND);
+		}
+
+		try {
+			$mailAccount = $this->accountService->find($this->userId, $accountId);
+		} catch (ClientException $e) {
+			$this->logger->error("Mail account #$accountId not found", ['exception' => $e]);
+			return new DataResponse('Account not found.', Http::STATUS_NOT_FOUND);
+		}
+
+		if($fromEmail !== $mailAccount->getEmail()) {
+			try {
+				$alias = $this->aliasesService->findByAliasAndUserId($fromEmail, $this->userId);
+			} catch (DoesNotExistException $e) {
+				$this->logger->error("Alias $fromEmail for mail account $accountId not found", ['exception' => $e]);
+				// Cannot send from this email as it is not configured as an alias
+				return new DataResponse("Could not find alias $fromEmail. Please check the logs.", Http::STATUS_NOT_FOUND);
+			}
+		}
+
+		if(empty($to)) {
+			return new DataResponse('Recipients cannot be empty.', Http::STATUS_BAD_REQUEST);
+		}
+
+		$attachments = $this->request->getUploadedFile('attachments');
+		$numberOfAttachments = count($attachments['name'] ?? []);
+
+		$localAttachments = [];
+		$messageAttachments = [];
+		$attachmentErrors = false;
+		while ($numberOfAttachments > 0) {
```

Comment:
`for` loops do exist :laughing:

Reply:
The parameter needs to be decremented before using it as an accessor, so a while loop is ideal here.